plugins/modules/katello_content_view_version.py

```diff
@@ -151,126 +136,98 @@
-    elif 'version' in params_dict:
-        content_view_version = find_content_view_version(module, content_view, version=params_dict['version'], failsafe=True)
+    if 'current_lifecycle_environment' in entity_dict:
+        entity_dict['current_lifecycle_environment'] = module.find_resource_by_name(
+            'lifecycle_environments', name=entity_dict['current_lifecycle_environment'], params=scope)
+        search = "content_view_id={},environment_ids=[{}]".format(content_view['id'], entity_dict['current_lifecycle_environment']['id'])
+        content_view_version = module.find_resource('content_view_versions', search=search, thin=True)
+    elif 'version' in entity_dict:
+        search = "content_view_id={},version={}".format(content_view['id'], entity_dict['version'])
+        content_view_version = module.find_resource('content_view_versions', search=search, thin=True, failsafe=True)
     else:
         content_view_version = None
 
     changed = False
     le_changed = False
-    if state == 'present':
+    if module.desired_absent:
+        changed = module.ensure_entity_state('content_view_version', None, content_view_version, params=scope)
+    else:
         if content_view_version is None:
-            kwargs = dict(data=dict())
-            if 'description' in params_dict:
-                kwargs['data'].update(description=params_dict['description'])
-            if 'force_metadata_regeneration' in params_dict:
-                kwargs['data'].update(force_yum_metadata_regeneration=params_dict['force_metadata_regeneration'])
-            if 'version' in params_dict:
-                kwargs['data'].update(major=list(map(int, str(params_dict['version']).split('.')))[0])
-                kwargs['data'].update(minor=list(map(int, str(params_dict['version']).split('.')))[1])
-
-            response = content_view.publish(params_dict['synchronous'], **kwargs)
-            changed = True
-            content_view_version = ContentViewVersion(id=response['output']['content_view_version_id']).read()
-
-        if 'lifecycle_environments' in params_dict:
-            lifecycle_environments = find_lifecycle_environments(module, names=params_dict['lifecycle_environments'], organization=organization)
-            le_changed = promote_content_view_version(module, content_view_version, organization, lifecycle_environments, params_dict['synchronous'],
-                                                      force=params_dict['force'],
-                                                      force_yum_metadata_regeneration=params_dict['force_yum_metadata_regeneration'])
-    elif state == 'absent':
-        changed = naildown_entity_state(ContentViewVersion, dict(), content_view_version, state, module)
+            # Do a sanity check, whether we can perform this non-synchronous
+            if 'lifecycle_environments' in entity_dict and not entity_dict['synchronous']:
```

Comment:
we end up here when `content_view_version is None` which can be in the following cases:
1. the user did not pass neither `version` nor `current_lifecycle_environment`
2. the user *did* pass a `version`, but it was a not-yet-existing one

obviously, for 2 we need to talk to the API to find out that the version didn't exist yet and we need to publish it first.
but for 1 we can know that we need to do a publish before doing any API interaction.

I wonder if we should add a similar test directly after `module.connect()` and thus fail as early as possible?

```python
if 'lifecycle_environments' in entity_dict and 'version' not in entity_dict and 'current_lifecycle_environment' not in entity_dict and not entity_dict['synchronous']:
    module.fail_json(msg="Cannot perform non-blocking publishing and promoting in the same module call.")
```

Reply:
I think, i'll put this even before `module.connect()`.